### ISSUE TITLE
`juliac` fails to compile Blink.jl: `JULIA_HOME not defined`

### ISSUE THREAD

**Issue by NHDaly** [JuliaLang/PackageCompiler.jl#47]
`juliac` is failing when I try to compile a file that uses `Blink.jl`, complaining that `LoadError: UndefVarError: JULIA_HOME not defined`.

Here is the very simple file I was compiling:
```
# test/blink_test.jl

using Blink

Base.@ccallable function julia_main(args::Vector{String})::Cint
    win = Window(shell(), Dict(:width=>1100))
    return 0
end
```

And here's the failure:

```
  (11:17:03) $ julia ~/.julia/v0.6/PackageCompiler/juliac.jl -vaej test/blink_test.jl ~/.julia/v0.6/PackageCompiler/examples/program.c builddir
Julia program file:
  "/Users/daly/Documents/developer/code/deep/JuliaDeep/test/blink_test.jl"
C program file:
  "/Users/daly/.julia/v0.6/PackageCompiler/examples/program.c"
Build directory:
  "/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir"
Make build directory
Change to build directory
Build module image files ".ji" in directory "/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir":
  `/Applications/Julia-0.6.app/Contents/Resources/julia/bin/julia -Cnative -J/Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib --compile=yes --depwarn=yes --startup-file=no --precompiled=no --compilecache=no -e 'empty!(Base.LOAD_CACHE_PATH) # reset / remove any builtin paths
        push!(Base.LOAD_CACHE_PATH, abspath("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir")) # enable usage of precompiled files
        include("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/blink_test.jl") # include Julia program file
        empty!(Base.LOAD_CACHE_PATH) # reset / remove build-system-relative paths'`
Build object file "blink_test.o" in directory "/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir":
  `/Applications/Julia-0.6.app/Contents/Resources/julia/bin/julia -Cnative -J/Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib --compile=yes --depwarn=yes --startup-file=no --precompiled=no --compilecache=no --output-o /Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir/blink_test.o -e 'empty!(Base.LOAD_CACHE_PATH) # reset / remove any builtin paths
        push!(Base.LOAD_CACHE_PATH, abspath("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir")) # enable usage of precompiled files
        include("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/blink_test.jl") # include Julia program file
        empty!(Base.LOAD_CACHE_PATH) # reset / remove build-system-relative paths'`
ERROR: LoadError: LoadError: LoadError: LoadError: LoadError: UndefVarError: JULIA_HOME not defined
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:576
 [2] eval(::Module, ::Any) at ./boot.jl:235
 [3] _require(::Symbol) at ./loading.jl:490
 [4] require(::Symbol) at ./loading.jl:405
 [5] include_from_node1(::String) at ./loading.jl:576
 [6] include(::String) at ./sysimg.jl:14
 [7] include_from_node1(::String) at ./loading.jl:576
 [8] include(::String) at ./sysimg.jl:14
 [9] include_from_node1(::String) at ./loading.jl:576
 [10] eval(::Module, ::Any) at ./boot.jl:235
 [11] _require(::Symbol) at ./loading.jl:490
 [12] require(::Symbol) at ./loading.jl:405
 [13] include_from_node1(::String) at ./loading.jl:576
 [14] include(::String) at ./sysimg.jl:14
while loading /Users/daly/.julia/v0.6/BinDeps/src/BinDeps.jl, in expression starting on line 10
while loading /Users/daly/.julia/v0.6/Blink/src/AtomShell/install.jl, in expression starting on line 1
while loading /Users/daly/.julia/v0.6/Blink/src/AtomShell/AtomShell.jl, in expression starting on line 9
while loading /Users/daly/.julia/v0.6/Blink/src/Blink.jl, in expression starting on line 11
while loading /Users/daly/Documents/developer/code/deep/JuliaDeep/test/blink_test.jl, in expression starting on line 1
ERROR: LoadError: failed process: Process(`/Applications/Julia-0.6.app/Contents/Resources/julia/bin/julia -Cnative -J/Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib --compile=yes --depwarn=yes --startup-file=no --precompiled=no --compilecache=no --output-o /Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir/blink_test.o -e 'empty!(Base.LOAD_CACHE_PATH) # reset / remove any builtin paths
        push!(Base.LOAD_CACHE_PATH, abspath("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/builddir")) # enable usage of precompiled files
        include("/Users/daly/Documents/developer/code/deep/JuliaDeep/test/blink_test.jl") # include Julia program file
        empty!(Base.LOAD_CACHE_PATH) # reset / remove build-system-relative paths'`, ProcessExited(1)) [1]
Stacktrace:
 [1] pipeline_error(::Base.Process) at ./process.jl:682
 [2] run(::Cmd) at ./process.jl:651
 [3] build_object(::String, ::String, ::String, ::Bool, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void) at /Users/daly/.julia/v0.6/PackageCompiler/src/static_julia.jl:217
 [4] #static_julia#1(::String, ::String, ::String, ::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::Void, ::String, ::Void, ::PackageCompiler.#static_julia, ::String) at /Users/daly/.julia/v0.6/PackageCompiler/src/static_julia.jl:127
 [5] (::PackageCompiler.#kw##static_julia)(::Array{Any,1}, ::PackageCompiler.#static_julia, ::String) at ./<missing>:0
 [6] julia_main(::Array{String,1}) at /Users/daly/.julia/v0.6/PackageCompiler/juliac.jl:130
 [7] include_from_node1(::String) at /Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 [8] include(::String) at /Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 [9] process_options(::Base.JLOptions) at /Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 [10] _start() at /Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia/sys.dylib:?
while loading /Users/daly/.julia/v0.6/PackageCompiler/juliac.jl, in expression starting on line 136
  (11:18:23) $
```

**Comment by SimonDanisch** [comment]
Is this a regression?

**Comment by NHDaly** [comment]
Oops, sorry, I didn't mean to hit "submit" until I'd linked to https://github.com/JuliaLang/PackageCompiler.jl/pull/44. I remember reading about `JULIA_HOME` there...

**Comment by NHDaly** [comment]
> Is this a regression?
I'm not sure. I know that I was unable to build this same project in the past, but I believe this specific failure is new. I'm reverting to a few commits back and trying again, i'll let you know. Sorry i hit "submit" before i meant to!

**Comment by SimonDanisch** [comment]
No worries :D Maybe #44 could even fix it?

**Comment by NHDaly** [comment]
haha omg it takes so longgggggg to test. 😭

Yes it is a regression. That file builds and executes correctly at  6f6740e, but fails at HEAD, c07a075, after #44 was merged.

> No worries :D Maybe #44 could even fix it?

:( sadly, no i think #44 broke it.

**Comment by NHDaly** [comment]
Looking at #44 more, perhaps it only fixed `JULIA_HOME` for snooping, but not for static_compile?:
884a9d143f3

It looks like that commit (884a9d143f3) was intended to fix the same problem I'm seeing, but only touched `snooping.jl`.

**Comment by NHDaly** [comment]
Simply reverting 267dd329, the commit that set "force no-compilecache /precompile" on all builds, does indeed fix the regression.

So perhaps we need to somehow add those same lines (`Sys.__init__(); Base.early_init()`) to the files being compiled by `juliac` as well?

/cc @TsurHerman

**Comment by TsurHerman** [comment]
This is how I understand things:
currently in my local `master` branch of PackageCompiler (an also part of the bootstrapping PR I issued)
I have these lines:
```
push!(julia_cmd.exec, "--precompiled=yes")
push!(julia_cmd.exec, "--compilecache=no")
push!(julia_cmd.exec, "--history-file=no")
```

precompile=yes because it speeds things up obviously by using precompiled code of a package if
such code exists, but compilecache=no so it means that it will not try and generate the precompiled code if it is missing. 
This solves at least one bug with generating a sysimg containing GLVisualise.
another reason not to turn compilecache is that it is not needed in this run and invoking the compiler
to generate precompiled code as part of a run thats its output is .. a precompiled code  is asking for trouble.

About 
` (Sys.__init__(); Base.early_init())`
It is basically the minimum steps required to get `JULIA_HOME` defined, as I encountered that any package that use JULIA_HOME failed with errors similar to what you reported.

in bootstrap.jl I changed these lines to be:
`try JULIA_HOME;catch Sys.__init__();Base.early_init();end`
so if JULIA_HOME is defined don't do anything.

I still get the annoying 
WARNING: redefining constant JULIA_HOME

so I guess even with precompiled=yes it is still somehow missing at startup.
I'll check again

**Comment by TsurHerman** [comment]
A possible explanation is that `Blink` module contains in it some code 
like 
`const _home = JULIA_HOME`
And if you have a precompiled version ready and --precompile=yes then it will work for you without
adding the the said lines.
Try going back to a commit where your program compiles and delete the cache for Blink 
(delete `.cache` and `lib` folders) and see if the regression returns.

**Comment by NHDaly** [comment]
> Try going back to a commit where your program compiles and delete the cache for Blink
(delete `.cache` and `lib` folders) and see if the regression returns.

:( Sadly it still built, so I don't think that's the explanation. That made sense though -- it was a good idea.

I went back to 6f6740e, and then ran `rm -rf ~/.julia/.cache/Blink/` `rm -rf ~/.julia/lib/v0.6/Blink.ji`, but it still built correctly.

**Comment by TsurHerman** [comment]
So which switch is in charge of failing 
your build?
—precompile=yes or
—compilecache=yes ?

**Comment by TsurHerman** [comment]
I think that when you have both switches on , if the precompiled cache is missing .. then it will generate the precompilation in another process. 

Just to validate things put —compilecache=no delete the cache and rebuild , I think you will see the regression.

Anyway JULIA_HOME is missing as part of a —output-o run and needs to be added by force.

I still recommend using the 
Try JULIA_HOME catch ... 
line at the beginning of any compilation script that gets called with —output-o switch , and I do recommend —compilecache=false as it solves at least one weird bug

**Comment by lucatrv** [comment]
@TsurHerman with PR #44 both `--precompile=no` and `--compilecache=no` are always passed by `juliac.jl`, however [here](https://github.com/JuliaLang/PackageCompiler.jl/issues/47#issuecomment-378363557) you write that you use "`--precompile=yes` because it speeds things up". Can you please clarify?

If you agree for now I will comment out the following lines, at least until the local cache behavior is clarified:
```
command = `$julia_cmd -e $expr`
verbose && println("Build module image files \".ji\" in directory \"$builddir\":\n  $command")
run(command)
```

CC: @vtjnash

**Comment by TsurHerman** [comment]
@lucatrv , The first PR was an attempt to solve some issues with generetaing a new sysimg.
The issue was some segfault when invoking the compiler to generate a compilecache for some package (--compilecahe=yes) and then using it (--precompile-yes) as part of the run that its output is supposed to be the new sysimg.

So at first I just disabled all precompilation and cacheing as it is not strictly needed since it is just a speed optimisation (premature optimisation?) , this solved the segfault issue.

At a later stage I pinpointed the segfault to occur only when both --precompile==yes and compilecachae=yes , and it makes sense then to have precompiled=yes as it can speed things up, 
and compilecache=no so if a local cache s missing then the package is just imported regularly and not
first invoked in another process using --output-ji switch and then imported .. 

Another issue is that JULIA_HOME which some packages depend upon ,  is not defined when 
using --output-o or --output-ji switches.
So to sum up my recommendations:

Set:

—compilecache=no
--precompile=yes 

always 

and add these lines to the top of any script:
```
if VERSION < v"0.7"
    try
        JULIA_HOME
    catch
        warn("JULIA_HOME is not defined, initializing manually")
        Sys.__init__()
        Base.early_init()
        JULIA_HOME
    end
end
```

**Comment by TsurHerman** [comment]
@lucatrv --precompile=yes speeds up module loading, if precompiled cache is available

**Comment by NHDaly** [comment]
> and add these lines to the top of any script

Okay yeah, i can confirm that adding those lines to the top of the script did indeed fix it. This script will compile and run successfully:

```julia
# blink_test.jl

if VERSION < v"0.7"
    try
        JULIA_HOME
    catch
        warn("JULIA_HOME is not defined, initializing manually")
        Sys.__init__()
        Base.early_init()
        JULIA_HOME
    end
end

using Blink

Base.@ccallable function julia_main(args::Vector{String})::Cint
    win = Window(shell(), Dict(:width=>1100))
    return 0
end
```

-----

Can/should we modify `juliac.jl` to automatically execute those lines before compiling a script? We could just add those lines before `include`ing the script here?:
https://github.com/JuliaLang/PackageCompiler.jl/blob/26b88d41d158026fc93f04efab7755af3088d3a2/src/static_julia.jl#L208-L210

```
        push!(Base.LOAD_CACHE_PATH, abspath(\"$builddir_esc\")) # enable usage of precompiled files
        try
            JULIA_HOME
        catch
            warn("JULIA_HOME is not defined, initializing manually")
            Sys.__init__();  Base.early_init();  JULIA_HOME
        end
        include(\"$juliaprog\") # include Julia program file
        empty!(Base.LOAD_CACHE_PATH) # reset / remove build-system-relative paths"
```

**Comment by NHDaly** [comment]
> Set:
>   —compilecache=no
>   --precompile=yes
> always

If I understand you correctly, we should revert this part of #44 then:
https://github.com/JuliaLang/PackageCompiler.jl/blob/26b88d41d158026fc93f04efab7755af3088d3a2/src/static_julia.jl#L182
and switch back to only conditionally setting it when using custom `cpu_target`?

I opened this PR to do that if you think it's right? :)
https://github.com/JuliaLang/PackageCompiler.jl/pull/50

**Comment by TsurHerman** [comment]
> and switch back to only conditionally setting it when using custom cpu_target?

Ah yes indeed I think you are right, revert it to using 
`push!(julia_cmd.exec, "--precompiled=no"); ` and it should be automatically no if compiling for a different cpu_target (how likely is this use case?) 
and yes if compiling for the `native` `cpu_target`

**Comment by NHDaly** [comment]
> how likely is this use case?
I think it will end up being fairly common! To distribute anything you compile, you have to change the cpu_target to accommodate other machines. :)

Okay great, thanks for your help, @TsurHerman! Let's have #50 submitted then.

**Comment by lucatrv** [comment]
I am still confused about the proper `--compilecache` and `--precompile` configuration for static compilation. The following lines:
```
expr = "Base.init_depot_path() # initialize package depots
        Base.init_load_path() # initialize location of site-packages
        empty!(Base.LOAD_CACHE_PATH) # reset / remove any builtin paths
        push!(Base.LOAD_CACHE_PATH, abspath(\"$builddir_esc\")) # enable usage of precompiled files
        include(\"$juliaprog\") # include Julia program file
        empty!(Base.LOAD_CACHE_PATH) # reset / remove build-system-relative paths"
```
initialize the location for modules' precompilation in the local build directory.
So I wonder how could `--precompile=yes` speed up compilation if used with `--compilecache=no`, since modules `.ji` image files would not be created in the build directory.
Moreover I wonder if this is just due to a Julia bug, and so if we should open an issue on the Julia repository instead of changing `static_julia.jl`.
But maybe I am just missing information, please let me know.
Thanks

**Comment by TsurHerman** [comment]
I guess `compilecache=no` `precompile=yes` makes sense only when compiling for the current cpu_target and when using the standard location for precompiled modules .. currently this only happens when building a better sysimg (Package Compiling).

In all other cases as you say , precompilation data is not available .. so in that case there is no point in precompiling a cache and then importing it , might as well just import it as it encompasses in it the precompilation and in a more stable way( no compiling in one process and importing it in another)

So I think it is safe to leave `compilecache=no` `precompile=yes` and to read it like
In the rare case where miraculously there exist a valid precompiled data .. by all means go ahead and use it.

**Comment by lucatrv** [comment]
Ok, we could do that for now. I am actually also working on adding options to customize `compilecache` `precompile` from `juliac.jl`, and for now I can keep default `compilecache=no` and `precompile=yes`.
However this seems to me due to a Julia bug with static compilation, so if you agree we should also open an issue on Julia repository. In future once the bug is fixed we can set default `compilecache=yes` and `precompile=yes` (in my opinion it should also work with a custom `cpu_target`, by rebuilding the cache).

**Comment by NHDaly** [comment]
> However this seems to me due to a Julia bug with static compilation

Do you have a better sense of whether this is indeed a Julia bug? Could you clarify for me what you think the bug is? I still don't think I understand.

----------------

Regarding the missing `JULIA_HOME` definition:
Can we add the above fix (manually defining it) to the julia code executed as part of `juliac.jl`? It seems silly for users to have to put that section in their code.

**Comment by NHDaly** [comment]
> Can we add the above fix (manually defining it) to the julia code executed as part of juliac.jl? It seems silly for users to have to put that section in their code.

Okay i opened a PR for that. we can move the conversation about that specific question there.
https://github.com/JuliaLang/PackageCompiler.jl/pull/55

**Comment by lucatrv** [comment]
@NHDaly sorry for late answer but I was traveling.

> Do you have a better sense of whether this is indeed a Julia bug? Could you clarify for me what you think the bug is? I still don't think I understand.

If I understand correctly, `--compilecache=yes` with static compilation gives the following two issues:

1. The segfault described above.
2. Issues with custom `cpu_target`, because packages precompiled for the wrong target would be imported. In this case I think `julia` should recognize the different target, and automatically recompile the packages for it, do you agree?

**Comment by NHDaly** [comment]
Ah, I see. Yes, certainly the segfault seems like a bug and should be reported.

Regarding the `cpu_target`, I can see it either way. It would be nice if julia just did the Right Thing, but it also seems acceptable to me that those flags are incompatible. I think either way would be reasonable. But you should definitely open an issue to see what others think!

**Comment by lucatrv** [comment]
@TsurHerman I guess that `Sys.__init__(); Base.early_init();` are needed also in Julia v0.7 to define `JULIA_BINDIR`, can you please verify?

**Comment by NHDaly** [comment]
Can we close this issue now? Everything seems to be working here now! :)

**Comment by lucatrv** [comment]
I'm keeping this open because:

1. I still have to open an issue to see what Julia developers think about the need to set `--compilecache="no"` and `--precompiled="no"` (sorry my fault, I will do it but it's hard for me to find the needed spare time, otherwise please feel free to open it yourself...)
2. I still see the annoying message `WARNING: redefining constant JULIA_HOME`

**Comment by lucatrv** [comment]
Answered [here](https://github.com/JuliaLang/PackageCompiler.jl/pull/116#issuecomment-424841153).
